File: heston_model_simulation/broadie-kaya-scheme/cpp_version/src/bessel.cpp
/**
 * @brief This script is used to calculate the value of the Bessel's function.
 * More details on the same can be found below:
 *
 * https://en.wikipedia.org/wiki/Bessel_function
 *
 * @author Harsh Parikh
 */

#include <iostream>
#include <cmath>
#include <complex>
#include "../helpers/gamma.hpp"
#include "../helpers/bessel.hpp"

std::complex<double> PowerScheme(std::complex<double> z,
                                 double alpha,
                                 double tolerance = 1e-8,
                                 int num_iterations = 100,
                                 bool log_space = true)
{
    // First term: k = 0
    std::complex<double> term = 1.0 / GammaFunction(alpha + 1.0);
    std::complex<double> sum = term;

    // The following formula is calculated using the recurrence method.
    /*
    term_k/term_{k-1} = 0.25 * z^2 * 1/(k(alpha + k))

    More details can be found in `CALCULATIONS.md` file.
    */
    for (int k = 1; k < num_iterations; ++k)
    {
        // term_k = term_{k-1} * (z^2/4) / (k * (alpha + k))
        term *= (0.25 * z * z) / (static_cast<double>(k) * (alpha + static_cast<double>(k)));
        sum += term;

        /*
        Here, the overall logic is as follows:

        If the value of the term is `x`, we need to ensure that the value
        is contributing a decent proportion to the overall sum. Hence, the
        proportion of x/sum should be greater than the tolerance in order to
        ensure a significant contribution.
        */
        if (std::abs(term) < tolerance * std::abs(sum))
        {
            // Returning the logarithmic variant
            return log_space ? (alpha * std::log(0.5 * z) + sum) : std::pow(0.5 * z, alpha) * sum;
        }
    }
    throw std::runtime_error("PowerScheme: Convergence not achieved within iteration limit");
}

std::complex<double> AsymptoticExpansion(std::complex<double> z,
                                         double alpha,
                                         double tolerance = 1e-8,
                                         int num_iterations = 100,
                                         bool log_space = true)
{
    // First term, k=0
    std::complex<double> term(1.0, 0.0);
    std::complex<double> sum = term;
    double prev_abs = std::abs(term);

    for (int k = 1; k < num_iterations; ++k)
    {
        /*
        The asymptotic series for I_alpha(z) is an alternating series:

            I_alpha(z) ~ e^z / sqrt(2*pi*z) * sum_{k=0}^inf (-1)^k * a_k(alpha) / z^k

        where:

            a_k(alpha) = prod_{j=1}^{k} (4*alpha^2 - (2j-1)^2) / (k! * 8^k)

        The recurrence for the term ratio is:

            term_k / term_{k-1} = -(4*alpha^2 - (2k-1)^2) / (8 * k * z)

        Expanding with mk = (2k - 1):

            -(4*alpha^2 - mk^2) / (8 * k * z)

        The negation produces the required alternating signs: without it the
        series accumulates same-sign terms and converges to the wrong value.
        */
        double mk = 2.0 * static_cast<double>(k) - 1.0;
        term *= -(4.0 * alpha * alpha - mk * mk) /
                (8.0 * static_cast<double>(k) * z);

        double curr_abs = std::abs(term);

        /*
        Minimum-term stopping rule for asymptotic series:

        Unlike a convergent power series, this asymptotic expansion eventually
        diverges for any fixed z. Once the term magnitude starts growing again,
        adding further terms only increases the error. We therefore stop as
        soon as the current term is larger than the previous one, keeping only
        the partial sum accumulated so far (up to and not including the
        diverging term).

        This rule is applied BEFORE adding the current term to the sum so that
        only the well-behaved portion of the series is retained.
        */
        if (curr_abs > prev_abs)
        {
            break;
        }

        sum += term;

        /*
        Here, the overall logic is as follows:

        If the value of the term is `x`, we need to ensure that the value
        is contributing a decent proportion to the overall sum. Hence, the
        proportion of x/sum should be greater than the tolerance in order to
        ensure a significant contribution.
        */
        if (curr_abs < tolerance * std::abs(sum))
        {
            break;
        }

        prev_abs = curr_abs;
    }

    // For I_alpha(z), the leading factor is e^z / sqrt(2*pi*z)

    // Returning the logarithmic value to avoid overflow for large z
    return log_space ? z - 0.5 * std::log(2.0 * M_PI * z) + std::log(sum)
                     : (std::exp(z) / std::sqrt(2.0 * M_PI * z)) * sum;
}

std::complex<double> ModifiedBessel(std::complex<double> z,
                                    double alpha,
                                    int num_iterations = 100,
                                    double tolerance = 1e-10,
                                    double threshold = 10.0,
                                    bool log_space = true)
{
    // NOTE: Negative integer orders (alpha < 0, integer) currently return
    // inf/nan due to a known bug in the symmetry redirect.
    // This does not affect Broadie-Kaya usage where alpha = d/2 - 1 > 0
    // for valid Heston parameters. Deferred for future fix.

    if (log_space && (alpha == 0))
    {
        throw std::domain_error("The log space cannot exist when alpha=0.");
    }

    // Handling negative integer orders using symmetry
    if (alpha < 0.0 && std::abs(std::round(alpha) - alpha) < 1e-12)
    {
        // For negative integers, I_{-n}(z) = I_n(z)
        return ModifiedBessel(z, -alpha, num_iterations, tolerance, threshold);
    }

    // Base cases for zero values
    // For real order α:
    // I_0(0) = 1 (by analytic continuation)
    // I_α(0) = 0 for α > 0
    // I_α(0) diverges (→ ∞) for α < 0

    if (z == std::complex<double>(0.0, 0.0))
    {
        if (alpha < 0.0)
        {
            throw std::domain_error("I_alpha(0) diverges to infinity for alpha < 0");
        }
        if (alpha == 0.0)
        {
            return log_space ? std::complex<double>(0.0, 0.0) : std::complex<double>(0.0, 0.0); // I_0(0) = 1
        }
        // alpha > 0
        // Here, the value of log(0) would be -inf. Hence, the code will raise a warning first if
        // the log space is activated.
        return std::complex<double>(0.0, 0.0);
    }

    // Choose method based on threshold (not tolerance!)
    if (std::abs(z) <= threshold)
    {
        return PowerScheme(z, alpha, tolerance, num_iterations, log_space);
    }
    else
    {
        return AsymptoticExpansion(z, alpha, tolerance, num_iterations, log_space);
    }
}

File: heston_model_simulation/broadie-kaya-scheme/python_version/helpers/bessel.py
"""
bessel.py
---------
Modified Bessel function of the first kind I_α(z) for complex arguments.

Two numerical schemes are implemented and selected based on |z|:
  - PowerScheme       : convergent power series, accurate for small |z|
  - AsymptoticExpansion: asymptotic series with minimum-term stopping rule,
                         accurate for large |z|

Both schemes support log-space evaluation to avoid overflow for large |z|,
which is critical for the Broadie-Kaya simulation where Bessel arguments
can be large.

Reference:
    https://en.wikipedia.org/wiki/Bessel_function
    Abramowitz & Stegun, Chapter 9

Author: Harsh Parikh
"""

import numpy as np
from gamma import gamma_real


def _power_scheme(
    z: complex,
    alpha: float,
    tolerance: float = 1e-8,
    num_iterations: int = 100,
    log_space: bool = True,
) -> complex:
    """
    Evaluate I_α(z) via convergent power series (small |z| regime).

    The series is:
        I_α(z) = (z/2)^α * Σ_{k=0}^∞  (z²/4)^k / (k! * Γ(α+k+1))

    The k-th term ratio recurrence is:
        term_k / term_{k-1} = (z²/4) / (k * (α+k))

    In log-space, returns:
        α*log(z/2) + log(Σ term_k / Γ(α+1))

    Parameters
    ----------
    z : complex
        Argument of I_α.
    alpha : float
        Order of the Bessel function.
    tolerance : float
        Convergence tolerance: stop when |term| < tolerance * |sum|.
    num_iterations : int
        Maximum number of series terms.
    log_space : bool
        If True, return log(I_α(z)) to avoid overflow.

    Returns
    -------
    complex
        log(I_α(z)) if log_space=True, else I_α(z).

    Raises
    ------
    RuntimeError
        If the series does not converge within num_iterations.
    """
    # First term k=0: 1 / Γ(α+1)
    term = np.complex64(1.0 / gamma_real(alpha + 1.0))
    total = term

    for k in range(1, num_iterations):
        # Recurrence: term_k = term_{k-1} * (z²/4) / (k*(α+k))
        term *= (0.25 * z * z) / (k * (alpha + k))
        total += term

        if abs(term) < tolerance * abs(total):
            if log_space:
                return alpha * np.log(0.5 * z) + np.log(total)
            else:
                return (0.5 * z) ** alpha * total

    raise RuntimeError(
        f"PowerScheme: convergence not achieved within {num_iterations} iterations "
        f"(|z|={abs(z):.4f}, alpha={alpha:.4f})"
    )


def _asymptotic_expansion(
    z: np.complex64,
    alpha: float,
    tolerance: float = 1e-8,
    num_iterations: int = 100,
    log_space: bool = True,
) -> np.complex64:
    """
    Evaluate I_α(z) via asymptotic expansion (large |z| regime).

    The asymptotic series is:
        I_α(z) ~ e^z / √(2πz) * Σ_{k=0}^∞ (-1)^k * a_k(α) / z^k

    where the term ratio recurrence is:
        term_k / term_{k-1} = -(4α² - (2k-1)²) / (8kz)

    Since this is an asymptotic (divergent) series, the minimum-term
    stopping rule is applied: stop as soon as |term_k| > |term_{k-1}|,
    retaining only the well-behaved partial sum.

    In log-space, returns:
        z - 0.5*log(2πz) + log(Σ terms)

    Parameters
    ----------
    z : np.complex64
        Argument of I_α (should satisfy |z| >> 1 for accuracy).
    alpha : float
        Order of the Bessel function.
    tolerance : float
        Secondary convergence tolerance.
    num_iterations : int
        Maximum number of asymptotic terms.
    log_space : bool
        If True, return log(I_α(z)) to avoid overflow.

    Returns
    -------
    np.complex64
        log(I_α(z)) if log_space=True, else I_α(z).
    """
    term = np.complex64(1.0)
    total = term
    prev_abs = abs(term)

    for k in range(1, num_iterations):
        # term_k/term_{k-1} = -(4α²-(2k-1)²) / (8kz)
        mk = 2.0 * k - 1.0
        term *= -(4.0 * alpha * alpha - mk * mk) / (8.0 * k * z)
        curr_abs = abs(term)

        # Minimum-term stopping rule: divergence detected, stop before adding
        if curr_abs > prev_abs:
            break

        total += term

        if curr_abs < tolerance * abs(total):
            break

        prev_abs = curr_abs

    if log_space:
        return z - 0.5 * np.log(2.0 * np.pi * z) + np.log(total)
    else:
        return (np.exp(z) / np.sqrt(2.0 * np.pi * z)) * total


def modified_bessel(
    z: np.complex64,
    alpha: float,
    num_iterations: int = 100,
    tolerance: float = 1e-10,
    threshold: float = 10.0,
    log_space: bool = True,
) -> np.complex64:
    """
    Evaluate the modified Bessel function of the first kind I_α(z).

    Selects between PowerScheme (|z| ≤ threshold) and AsymptoticExpansion
    (|z| > threshold) automatically. Log-space evaluation is recommended
    for large |z| to avoid floating-point overflow.

    Note
    ----
    Negative integer orders are handled via the symmetry I_{-n}(z) = I_n(z).
    The case alpha=0 with log_space=True raises an error (log(I_0) at z=0
    is undefined; use log_space=False for alpha=0).

    Parameters
    ----------
    z : np.complex64
        Argument of I_α. For Broadie-Kaya, this is a positive real number
        scaled by variance parameters.
    alpha : float
        Order of the Bessel function. In Heston: alpha = d/2 - 1 where
        d = 4κθ/σ² is the degrees of freedom of the CIR process.
    num_iterations : int
        Maximum iterations for either series.
    tolerance : float
        Convergence tolerance.
    threshold : float
        |z| threshold for switching from power series to asymptotic expansion.
    log_space : bool
        If True, return log(I_α(z)). Strongly recommended for large |z|.

    Returns
    -------
    np.complex64
        log(I_α(z)) if log_space=True, else I_α(z).

    Raises
    ------
    ValueError
        If alpha=0 and log_space=True, or if z=0 with alpha < 0.

    Examples
    --------
    >>> # Compare with scipy for validation
    >>> from scipy.special import iv
    >>> z, alpha = 5.0, 1.0
    >>> result = modified_bessel(z, alpha, log_space=False).real
    >>> expected = iv(alpha, z)
    >>> abs(result - expected) < 1e-6
    True
    """
    z = np.complex64(z)

    if log_space and alpha == 0:
        raise ValueError("log_space=True is undefined at alpha=0 when z=0.")

    # Symmetry: I_{-n}(z) = I_n(z) for integer n
    if alpha < 0 and abs(round(alpha) - alpha) < 1e-12:
        return modified_bessel(
            z, -alpha, num_iterations, tolerance, threshold, log_space
        )

    # Base cases at z=0
    if z == 0.0:
        if alpha < 0:
            raise ValueError("I_α(0) diverges for α < 0")
        # I_0(0) = 1 → log = 0; I_α(0) = 0 for α > 0 → log = -inf
        return np.complex64(0.0)

    if abs(z) <= threshold:
        return _power_scheme(z, alpha, tolerance, num_iterations, log_space)
    else:
        return _asymptotic_expansion(z, alpha, tolerance, num_iterations, log_space)
